### PR TITLE
[Sprint 49] XD-2992 Support multiple topics for kafka source

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- Allow comma separated list of topics as module option to kafka source module
 - Restrict initialOffset and explicit partition list only for single topic
   - Add validation that checks these options are valid only if single topic is configured
 - Modify `KafkaPartitionAllocator`
   - Remove ZooKeeper dependency to persist the partitions for the kafka source
   - Evenly distribute the kafka partitions (all or specific) to the requested module instances
 - Update test

### CODE REVIEW

**Comment by ilayaperumalg** [comment]
Created https://jira.spring.io/browse/XD-3035 to handle the deletion of offsets upon stream destroy.

File: spring-xd-shell/src/test/java/org/springframework/xd/shell/command/KafkaSourceSinkTests.java [hunk 1]
```diff
@@ -81,16 +94,21 @@ public void deleteTopic() {
...
 		// create stream with kafka source
 		final CounterSink counter = metrics().newCounterSink();
-		stream().create(generateStreamName(), "kafka --topic='%s' --zkconnect=%s --outputType=text/plain | " +
+		stream().create(generateStreamName(), "kafka --topics='%s,%s' --zkconnect=%s --outputType=text/plain | " +
 				"filter --expression=payload.toString().contains('%s') | %s",
-				topicToUse, kafkaTestSupport.getZkConnectString(), stringToPost, counter );
-		httpSource.ensureReady().postData(stringToPost);
+				topic1ToUse, topic2ToUse, kafkaTestSupport.getZkConnectString(), stringToPost, counter );
+		httpSource1.ensureReady().postData(stringToPost);
+		httpSource1.ensureReady().postData(stringToPost);
```
Comment (mbogoevici) [inline spring-xd-shell/src/test/java/org/springframework/xd/shell/command/KafkaSourceSinkTests.java]: the test posts data to the same http source - should be `httpSource2` instead

File: extensions/spring-xd-extension-kafka/src/main/java/org/springframework/integration/x/kafka/KafkaSourceModuleOptionsMetadata.java [hunk 2]
```diff
@@ -31,7 +36,7 @@
 @Mixin({KafkaZKOptionMixin.class, KafkaConsumerOptionsMixin.class, KafkaOffsetTopicOptionsMixin.class})
 public class KafkaSourceModuleOptionsMetadata implements ProfileNamesProvider {
 
-	private String topic = ModulePlaceholders.XD_STREAM_NAME;
+	private String topics = ModulePlaceholders.XD_STREAM_NAME;
```
Comment (mbogoevici) [inline extensions/spring-xd-extension-kafka/src/main/java/org/springframework/integration/x/kafka/KafkaSourceModuleOptionsMetadata.java]: I'm wondering if we shouldn't support both options (`topic`) and (`topics`) and allow only one to be set - for backwards compatibility - e.g. for existing deployments, etc. Granted, this adds some complexity ...

File: spring-xd-shell/src/test/java/org/springframework/xd/shell/command/KafkaSourceSinkTests.java (hunk 1 above)
Comment (mbogoevici) [inline spring-xd-shell/src/test/java/org/springframework/xd/shell/command/KafkaSourceSinkTests.java]: this is sort of minor, can be fixed on merge.

File: extensions/spring-xd-extension-kafka/src/main/java/org/springframework/integration/x/kafka/KafkaSourceModuleOptionsMetadata.java (hunk 2 above)
Comment (ilayaperumalg) [inline extensions/spring-xd-extension-kafka/src/main/java/org/springframework/integration/x/kafka/KafkaSourceModuleOptionsMetadata.java]: We can have both options `topic` (at least for backward compatibility) and `topics` with a validation condition saying only one can be set at a given time.

**Comment by ilayaperumalg** [comment]
Pushed the changes that makes both `topic` and `topics` available as module options while keeping them  mutually exclusive.

**Comment by mbogoevici** [comment]
Rebased, merged w/https://github.com/spring-projects/spring-xd/commit/2ffeb96367fb78981ac0ceedd64495c6e4c79ec5

**Pull request closed without merging**